### PR TITLE
fix: Solve the problem of final consistency when deleting

### LINKED ISSUE #118: feat: Support retries in upstream definition

### PR DESCRIPTION
**Pull request by gxthrj**
Solve the problem of final consistency when deleting.
Support retry when remove upstream or service failed.
Fix #118

### CODE REVIEW

File: pkg/apisix/resource.go [hunk 1]
```diff
@@ -116,6 +116,7 @@ func (i *item) upstream(clusterName string) (*v1.Upstream, error) {
 	}
 
 	var ups upstreamItem
+	log.Info(string(i.Value))
```
Comment (tokers) [inline pkg/apisix/resource.go:119]: It already logs `i.Value` in line 112.

File: pkg/ingress/controller/apisix_upstream.go [hunk 2]
```diff
@@ -112,30 +112,46 @@ func (c *ApisixUpstreamController) processNextWorkItem() bool {
...
+			log.Warnf("Upstram %s has been covered when retry", sqo.Key)
 			return nil
 		}
-		runtime.HandleError(fmt.Errorf("failed to list apisixUpstream %s/%s", key, err.Error()))
-		return err
+	} else {
+		apisixUpstreamYaml, err = c.apisixUpstreamList.ApisixUpstreams(namespace).Get(name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				log.Infof("apisixUpstraem %s is removed", sqo.Key)
```
Comment (tokers) [inline pkg/ingress/controller/apisix_upstream.go:132]: Type: `apisixUpstraem` => `apisixUpstream`

File: pkg/ingress/controller/apisix_upstream.go [hunk 3]
```diff
@@ -112,30 +112,46 @@ func (c *ApisixUpstreamController) processNextWorkItem() bool {
...
+				return nil
+			}
+			runtime.HandleError(fmt.Errorf("failed to list apisixService %s/%s", sqo.Key, err.Error()))
```
Comment (tokers) [inline pkg/ingress/controller/apisix_upstream.go:135]: Type: here we are listing `ApisixUpstream`.

File: pkg/ingress/controller/apisix_upstream.go [hunk 4]
```diff
@@ -145,25 +161,45 @@ func (c *ApisixUpstreamController) addFunc(obj interface{}) {
...
-	oldRoute := oldObj.(*apisixV1.ApisixUpstream)
-	newRoute := newObj.(*apisixV1.ApisixUpstream)
-	if oldRoute.ResourceVersion == newRoute.ResourceVersion {
+	oldUpstream := oldObj.(*apisixV1.ApisixUpstream)
+	newUpstream := newObj.(*apisixV1.ApisixUpstream)
+	if oldUpstream.ResourceVersion >= newUpstream.ResourceVersion {
 		return
 	}
-	c.addFunc(newObj)
+	var key string
```
Comment (tokers) [inline pkg/ingress/controller/apisix_upstream.go:174]: Style:

```go
var (
    key string
    err  error
)
```

File: test/e2e/scaffold/crd.go [hunk 5]
```diff
@@ -22,14 +22,13 @@ import (
 	"strconv"
 	"time"
 
+	"github.com/api7/ingress-controller/pkg/apisix"
```
Comment (tokers) [inline test/e2e/scaffold/crd.go:25]: Style: put these lines at the end of import block.

File: test/e2e/scaffold/ingress.go [hunk 6]
```diff
@@ -77,7 +77,7 @@ spec:
             tcpSocket:
               port: 8080
             timeoutSeconds: 2
-          image: "apache/apisix-ingress-controller:dev"
+          image: "apisix-ingress-controller:dev"
```
Comment (tokers) [inline test/e2e/scaffold/ingress.go:80]: Why change it again?

File: pkg/apisix/resource.go (hunk 1 above)
Comment (gxthrj) [reply pkg/apisix/resource.go:119]: Removed

File: pkg/ingress/controller/apisix_upstream.go (hunk 2 above)
Comment (gxthrj) [reply pkg/ingress/controller/apisix_upstream.go:132]: Done

File: pkg/ingress/controller/apisix_upstream.go (hunk 3 above)
Comment (gxthrj) [reply pkg/ingress/controller/apisix_upstream.go:135]: Done

File: pkg/ingress/controller/apisix_upstream.go (hunk 4 above)
Comment (gxthrj) [reply pkg/ingress/controller/apisix_upstream.go:174]: Done

File: test/e2e/scaffold/crd.go (hunk 5 above)
Comment (gxthrj) [reply test/e2e/scaffold/crd.go:25]: In the test mod , `github.com/api7/ingress-controller` is an externally imported package.
I used fmt tool for me to deal with group of imports.

File: test/e2e/scaffold/ingress.go (hunk 6 above)
Comment (gxthrj) [reply test/e2e/scaffold/ingress.go]: Done

File: test/e2e/scaffold/crd.go (hunk 5 above)
Comment (tokers) [reply test/e2e/scaffold/crd.go:25]: My bad.

**Review by tokers — APPROVED** [review]

**Pull request merged**